### DISCUSSION TITLE
DemonSet 方式在 ACK 中部署 ilogtail，删除 Pod 时出现 Terminating ，一直无法删除，只能强制删除

### DISCUSSION THREAD

**Discussion by hanzoi** [alibaba/ilogtail#511] [Help] [answered]
kubelet 日志如下：
<img width="1016" alt="企业微信截图_566bf17f-7125-4cc5-95f8-95a2b85c299d" src="https://user-images.githubusercontent.com/61114924/200220019-745b7a8e-7486-4b9c-82a8-5d723ea7d1fb.png">
<img width="1004" alt="企业微信截图_8b31e246-73f5-4441-9b90-502f69e5cd52" src="https://user-images.githubusercontent.com/61114924/200220030-fdad5102-1244-4ccd-b48b-abb4dbec498c.png">
<img width="1004" alt="企业微信截图_168e1696-7db0-4a4f-9185-ae95e26ce7b8" src="https://user-images.githubusercontent.com/61114924/200220040-7b4f1404-d982-426a-9527-4411a9c872ad.png">

**Comment by hanzoi** [comment]


**Reply by hanzoi** [reply]
再补充一下，目前直接使用 crictl rm 也无法删除容器，报错如下：
time="2022-11-07T13:51:50.382753914+08:00" level=warning msg="failed to cleanup rootfs mount" error="failed to unmount target /run/containerd/io.containerd.runtime.v2.task/k8s.io/ae376aebdb60f228a16c493720d9b31bb5d76679cbaf201d8207f0820f9b5788/rootfs: device or resource busy" runtime=io.containerd.runc.v2

os：Alibaba Cloud Linux  2.1903 LTS 64位

尝试设置 fs.may_detach_mounts=1 ，报错为：
![image](https://user-images.githubusercontent.com/61114924/200446645-b6ac1c6a-f0c2-4aed-b0f9-39e92a9de96c.png)

**Reply by hanzoi** [reply]
<img width="1247" alt="image" src="https://user-images.githubusercontent.com/61114924/202079878-e01b5cba-2958-41d7-aaae-ca8c83092841.png">

相似的配置文件一个 ilogtail 大概会挂 8个，每个都负责采集不同目录下的日志

**Comment by yyuuttaaoo** [comment, marked as answer]
补充一些信息

1. 从lsof的情况看，无法退出的容器看不到ilogtail持有句柄   
![image](https://user-images.githubusercontent.com/1827594/202081927-3b275094-45ba-404c-a0b7-ee85eb683acf.png)
2. 怀疑是CentOS的bug，尝试设置 fs.may_detach_mounts=1，但上面帖子反馈报错。建议还能再尝试一下这个帖子的设置方法
  https://github.com/kubernetes-sigs/kubespray/issues/8622#issuecomment-1068238054
3. logtail支持在容器退出时，强制一定时间内释放句柄可以尝试
   https://help.aliyun.com/document_detail/32278.html
   docker_config_update_interval 这个可以改成1 会不会好点
   force_release_deleted_file_fd_timeout 这个改成0会立刻释放句柄
4. 对于ilogtail文件打开过多的问题，可以看看/proc/$ilogtail_pid/fd，具体持有什么文件句柄了
